### PR TITLE
Make model stateless and context stateful (llama_state)

### PR DESCRIPTION
**Pull request by didzis**
This pull request aims to solve the `llama_state` problem by making `llama_context` the new "`llama_state`". This approach is fully backwards compatible and has only few additions to the public API to enable separate loading of model and making multiple contexts from this single model instance. I converted the `llama_model` and `llama_vocab` members to references, now `llama_context` has a constructor that takes `model` and `vocab` reference as parameters,  I also added `const` modifier for more compiler checks (can be removed). For simplicity of the public API I moved the `vocab` under `llama_model`.

### CODE REVIEW

**Comment by bullno1** [comment]
Since `llama_vocab` is already rolled into `llama_model`, is `llama_context::vocab` even necessary?

Should all the code just use `ctx->model.vocab` anyway?

**Comment by ggerganov** [comment]
Ok, this is looking very good. Will take a deeper look later, also want to see more feedback from other people

**Comment by xaedes** [comment]
I like the idea to make `llama_context` the new `llama_state`.
For C API compatibility I would prefer pointers instead of references.
Makes it easier to create for example python wrappers for the `llama_context`.

File: llama.cpp [hunk 1]
```diff
@@ -218,24 +232,11 @@ struct llama_model {
...
-        token tok;
-        float score;
-    };
-
-    std::unordered_map<token, id> token_to_id;
-    std::vector<token_score> id_to_token;
-};
-
 struct llama_context {
+    llama_context(const llama_model& model, const llama_vocab& vocab) : model(model), vocab(vocab), t_load_us(model.t_load_us), t_start_us(model.t_start_us) {}
```
Comment (ggerganov) [inline llama.cpp:236]:
```suggestion
    llama_context(const llama_model & model, const llama_vocab & vocab) : model(model), vocab(vocab), t_load_us(model.t_load_us), t_start_us(model.t_start_us) {}
```

File: llama.cpp [hunk 2]
```diff
@@ -246,8 +247,16 @@ struct llama_context {
     int32_t n_eval   = 0; // number of eval calls
     int32_t n_p_eval = 0; // number of tokens in eval calls for the prompt (with batch size > 1)
 
-    llama_model model;
-    llama_vocab vocab;
+    const llama_model& model;
+    const llama_vocab& vocab;
```
Comment (ggerganov) [inline llama.cpp:251]:
```suggestion
    const llama_model & model;
    const llama_vocab & vocab;
```

File: llama.cpp [hunk 3]
```diff
@@ -2521,12 +2531,38 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
...
 //
 
-struct llama_context * llama_init_from_file(
+struct llama_model * llama_load_model_from_file(
                              const char * path_model,
             struct llama_context_params   params) {
     ggml_time_init();
 
-    llama_context * ctx = new llama_context;
+    llama_model *model = new llama_model;
```
Comment (ggerganov) [inline llama.cpp:2539]:
```suggestion
    llama_model * model = new llama_model;
```

File: llama.cpp [hunk 4]
```diff
@@ -2521,12 +2531,38 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
...
+
+    ggml_type memory_type = params.f16_kv ? GGML_TYPE_F16 : GGML_TYPE_F32;
+
+    if (!llama_model_load(path_model, *model, model->vocab, params.n_ctx, params.n_batch, params.n_gpu_layers,
+                params.main_gpu, params.tensor_split, memory_type, params.use_mmap, params.use_mlock,
+                params.vocab_only, params.progress_callback, params.progress_callback_user_data)) {
+        fprintf(stderr, "%s: failed to load model\n", __func__);
```
Comment (ggerganov) [inline llama.cpp:2546]: delete `model` here

File: llama.cpp [hunk 5]
```diff
@@ -2521,12 +2531,38 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
...
+
+void llama_free_model(struct llama_model * model) {
+    delete model;
+}
+
+struct llama_context * llama_new_context_with_model(
+                             struct llama_model * model,
+            struct llama_context_params   params) {
+
+    if(!model) {
```
Comment (ggerganov) [inline llama.cpp:2561]:
```suggestion
    if (!model) {
```

File: llama.cpp [hunk 6]
```diff
@@ -2646,11 +2690,9 @@ int llama_model_quantize(
     }
 }
 
-int llama_apply_lora_from_file_internal(struct llama_context * ctx, const char * path_lora, const char * path_base_model, int n_threads) {
+int llama_apply_lora_from_file_internal(const struct llama_model& model, const char * path_lora, const char * path_base_model, int n_threads) {
```
Comment (ggerganov) [inline llama.cpp:2693]:
```suggestion
int llama_apply_lora_from_file_internal(const struct llama_model & model, const char * path_lora, const char * path_base_model, int n_threads) {
```

**Review by ggerganov — APPROVED** [review]
I think we can merge after addressing the minor review comments

After merging, we should create a basic example of doing 2 or more generations in parallel with a single model

File: llama.cpp [hunk 7]
```diff
@@ -2629,7 +2657,23 @@ struct llama_context * llama_init_from_file(
     return ctx;
 }
 
+struct llama_context * llama_init_from_file(
```
Comment (howard0su) [inline llama.cpp:2660]: I vote to remove this. I prefer giving user one choice only, which is:
model = load_model
context = new_context_from_model.

Giving second choice of create context from model file name is not worth the code and confusioni.

Comment (howard0su) [reply llama.cpp:2776]: #ifdef __GNUC__
#define DEPRECATED(func) func __attribute__ ((deprecated))
#elif defined(_MSC_VER)
#define DEPRECATED(func) __declspec(deprecated) func
#else
DEPRECATED(func) func
#endif

**Review by ggerganov — APPROVED** [review]
🦙

File: examples/common.h [hunk 8]
```diff
@@ -95,7 +96,7 @@ std::vector<llama_token> llama_tokenize(struct llama_context * ctx, const std::s
 // Model utils
 //
 
-struct llama_context * llama_init_from_gpt_params(const gpt_params & params);
+std::tuple<struct llama_model *, struct llama_context *> llama_init_from_gpt_params(const gpt_params & params);
```
Comment (ggerganov) [inline examples/common.h:99]: An alternative would have been to avoid `std::tuple` and introduce helper `struct llama_init_results`

**Pull request merged**